drivers: watchdog: Add Raspberry Pi Pico watchdog driver

thedjnK opened the pull request:
This adds basic support for the watchdog timer on the RP2040 MCU and
Raspberry Pi Pico development board

Tested on a raspberry pi pico board with a modified hello world sample, board reboots after ~8 seconds

thedjnK committed:
drivers: watchdog: Add Raspberry Pi Pico watchdog driver

This adds basic support for the watchdog timer on the RP2040 MCU and
Raspberry Pi Pico development board

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented:
@yonsch could you give this a review?

yonsch commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,174 @@
+LOG_MODULE_REGISTER(wdt_rpi_pico, CONFIG_WDT_LOG_LEVEL);
+
+/* Maximum watchdog time is halved due to errata RP2040-E1 */
+#define RPI_2040_MAX_WDT_TIME (0xffffff / 2)
```
`RPI_PICO` would be a better prefix, since this driver is not just for RP2040. But anyway, I don't think you even need to namespace it, since the definition is not exported outside of this file.

thedjnK replied:
Updated to RPI_PICO

yonsch commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,174 @@
+
+/* Maximum watchdog time is halved due to errata RP2040-E1 */
+#define RPI_2040_MAX_WDT_TIME (0xffffff / 2)
+#define RPI_2020_WDT_TIME_MULTIPLCATION_FACTOR 2
```
MULTIPLCATION is a typo

thedjnK replied:
Fixed

yonsch commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,174 @@
+#define RPI_2020_WDT_TIME_MULTIPLCATION_FACTOR 2
+
+/* Watchdog requires a 1MHz clock source, divided from the crystal oscillator */
+#define RPI_2040_XTAL_FREQ_WDT_TICK_DIVISOR 1000000
```
So, this is the target frequency, right? Not the divisor. This is used to calculate the divider for the particular clock speed.

thedjnK replied:
It is both, the target frequency is 1MHz but the tick count is in cycles, so at a clock speed of 12MHz, this needs to be divided by 1M to get 12, hence the name, as the tick count is fed from the 1MHz timer source.

yonsch commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,174 @@
+		.xtal_frequency = DT_INST_PROP_BY_PHANDLE(idx, clocks, clock_frequency)            \
+	};                                                                                         \
+	static struct wdt_rpi_pico_data wdt_##idx##_data = {                                       \
+		.reset_type = WDT_FLAG_RESET_SOC,                                                  \
```
Shouldn't `reset_type` and `load` be part of the config, not the data?

thedjnK replied:
These are runtime settable when calling the `install_timeout` function so cannot be part of the const config

yonsch reviewed (commented):
I'm no expert on Zephyr's WD API nor the pico's WD, but here's what I could find

thedjnK committed:
tests: drivers: watchdog: Add RPI Pico support

Adds support for running the wdt test on the Raspberry Pi Pico by
increasing the window size and disabling timeouts.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
doc: guides: release notes: Add Raspberry Pi Pico watchdog support

The Raspberry Pi Pico watchdog is now supported, this adds a note to
the release notes with this information.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

petejohanson commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,170 @@
+	} else if (cfg->window.max > RPI_PICO_MAX_WDT_TIME) {
+		return -EINVAL;
+	} else if (cfg->callback != NULL) {
+		return -EINVAL;
```
Shouldn't this be:


```suggestion
		return -ENOTSUP;
```
?

thedjnK replied:
It should, corrected

petejohanson reviewed (commented):
Not super familiar with the WDT stuff, but a couple general items I noticed.

thedjnK commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,170 @@
+	if (data->reset_type == WDT_FLAG_RESET_SOC) {
+		hw_set_bits(&psm_hw->wdsel, PSM_WDSEL_BITS);
+	} else if (data->reset_type == WDT_FLAG_RESET_CPU_CORE) {
+		hw_set_bits(&psm_hw->wdsel, PSM_WDSEL_PROC0_BITS);
```
Added a todo

petejohanson commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -0,0 +1,171 @@
+
+#include <hardware/watchdog.h>
+#include <hardware/structs/psm.h>
+#include <drivers/watchdog.h>
+
+#include <logging/log.h>
```
Should be using the new `zephyr/` prefix for includes, e.g.:

```suggestion
#include <zephyr/drivers/watchdog.h>

#include <zephyr/logging/log.h>
```

thedjnK replied:
Fixed

petejohanson reviewed (commented):
One other minor nitpick I noticed.

petejohanson commented on drivers/watchdog/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2022, Jamie McCrae
+# SPDX-License-Identifier: Apache-2.0
+
```
DT_COMPAT_RASPBERRYPI_PICO_WATCHDOG := raspberrypi,pico-watchdog

thedjnK replied:
Fixed

petejohanson commented on drivers/watchdog/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,18 @@
+
+config WDT_RPI_PICO
+	bool "Raspberry Pi Pico Watchdog driver"
+	default y
```
```suggestion
	default $(dt_compat_enabled,$(DT_COMPAT_RASPBERRYPI_PICO_WATCHDOG))
```

thedjnK replied:
Fixed

yonsch reviewed (dismissed):
LGTM

martinjaeger reviewed (dismissed):
Don't have any experience with RPi Pico, but the watchdog code looks very clean and sound. I assume you have successfully tested it in hardware.

As we just passed feature freeze for v3.1 we'll need to wait till the release beginning of June for this to get merged.

thedjnK commented:
Yes, the basic wdt test passes on the device.

yonsch commented:
@thedjnK One thing I forgot - please add watchdog to the table of supported features in the board's doc

thedjnK committed:
boards: rpi_pico: Added Watchdog timer to supported feature list

Adds Watchdog timer (WDT) to the list of supported features for the
Raspberry Pi Pico board.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
samples: drivers: watchdog: Add support for Raspberry Pi Pico

This adds support for running the watchdog driver sample on a Raspberry
Pi Pico board.

Signed-off-by: Jamie McCrae <spam@helper3000.net>
